Fix: Handle if `RangeSelector` min and max args are equal

Lamparter opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! ⚠ -->

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Fixes #366**

<!-- Add the relevant issue number after the word "Fixes" mentioned above (for ex: `## Fixes #0000`) which will automatically close the issue once the PR is merged. -->

<!-- PRs without assigned issues will be closed unless minor changes to documentation/typos. -->

<!-- Add a brief overview here of the feature/bug & fix. -->

**PR Type: Bugfix**

---

This PR fixes an issue where DWM crashes if the arguments for both min and max of the range selector control are equal.

<!-- Please uncomment one or more options below that apply to this PR. -->
<!-- New features should come from Windows Community Toolkit Labs. If you're migrating a component from Labs, link to the approval comment here. -->

<!-- - Bugfix -->
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

When the Maximum and Minimum arguments of `RangeSelector` are equal, DWM crashes. This can be reproduced by even the Toolkit gallery app.

## What is the new behavior?

<!-- Describe how was this issue resolved or changed? -->

If the max and min args of the range selector are equal, an exception will be thrown.

## PR Checklist

Please check if your PR fulfills the following requirements: <!-- and remove the ones that are not applicable to the current PR -->

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [ ] Tested code with current supported SDKs
- [x] Contains **NO** breaking changes
<!--- [ ] New component
  - [ ] Documentation has been added
  - [ ] Sample in sample app has been added
  - [ ] Analyzers are passing for documentation and samples
  - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/CommunityToolkit/WindowsCommunityToolkit-design-assets)-->
<!--- [ ] Tests for the changes have been added (if applicable)-->
<!--- [ ] Header has been added to all new source files-->


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

<!--## Other information-->

<!-- Please add any other information that might be helpful to reviewers. -->

Lamparter committed:
Fix issue

sylveon commented:
Should probably change the gallery app to prevent this as well (with the fix, gallery app would just crash)

Arlodotexe commented:
> Should probably change the gallery app to prevent this as well (with the fix, gallery app would just crash)

The same would happen in any app. The issue is in the `RangeSelector` control, it's crashing because it threw an unhandled exception. This fix should be enough :)

Lamparter commented:
But don't you need to handle the exception in the WCT gallery app?

Arlodotexe commented:
@michael-hawker There seems to be some undefined behavior here. Should the control simply ignore when `Min == Max`, should it prevent that tick, or something else?

michael-hawker commented:
We should pattern off `Slider`'s behavior here for what we should do. In the base case there, setting min/max to the same value is allowed, and the value is just locked to those values as well:

![image](https://github.com/user-attachments/assets/876ff633-e04a-4584-a143-0f5424d8f60d)

So, I think the best expected behavior is the same here; we should honor that same underlying contract.

Therefore, we should understand what's not working in this scenario to enable that vs. just throwing a different exception.

This would also be worth doing anyway, as we should root-cause the underlying issue; as we should file a platform bug around the DWM crash, as that shouldn't happen regardless. (I wonder if this repros on WinUI 3 as well...)

@Lamparter is this still something you could/would be willing to help us with?

Lamparter commented:
Sure!

michael-hawker commented:
@Lamparter thanks! Let us know if you need any assistance and how we can help support you.

Once we understand the root cause of what's causing the underlying crash, we can open/raise a platform issue on the [WinUI repo](https://github.com/microsoft/microsoft-ui-xaml) too with a minimal repro without the toolkit maybe.

Then, we can hopefully just find a good workaround in our toolkit code here to prevent that scenario from occurring when the edge case bounds are setup, and add a test or two to help us ensure we don't regress in the future.

Arlodotexe committed:
Merge branch 'main' into dwm

niels9001 committed:
Merge branch 'main' into dwm

Arlodotexe committed:
Merge branch 'main' into dwm

Arlodotexe committed:
Remove Minimum/Maximum Epsilon delta property and usage, fixed issue where Maximum values less than Minimum would erroneously overwrite Minimum

Arlodotexe committed:
Fixed issue where Minimum values greater than Maximum would erroneously overwrite minimum, cleaned up duplicate logic
Amends c28bc00151f247142eda16cce5b9cc82249f3b6b

Arlodotexe committed:
Ensure RangeStart/RangeEnd stay on valid steps when constrained by each other

When RangeStart is set to a value that snaps beyond RangeEnd (or vice versa), the control now snaps to the nearest valid step that respects the constraint, rather than being clamped to a non-step value.

Problem: With StepFrequency=40, RangeStart=40, RangeEnd=60, setting RangeStart=60 would:
1. Snap 60 → 80 (nearest step via Math.Round)
2. VerifyValues() clamps 80 → 60 (to not exceed RangeEnd)
3. Result: RangeStart=60, which is not on a valid step (valid steps: 0, 40, 80, 100)

Solution: RangeMinToStepFrequency and RangeMaxToStepFrequency now handle the range constraint directly after step snapping:
• RangeMinToStepFrequency: If snapped value exceeds RangeEnd, snap down to previous valid step using Math.Floor
• RangeMaxToStepFrequency: If snapped value is below RangeStart, snap up to next valid step using Math.Ceiling

This keeps MoveToStepFrequency as a pure step-snapping utility, while the range constraint logic lives in the methods that understand the relationship between the two range values.

Note: The interaction between VerifyValues() (bounds clamping) and step frequency snapping remains complex and should be refactored in a future PR to clarify responsibilities.

Arlodotexe committed:
Fixed order of operations for bound correction in VerifyValues

Arlodotexe committed:
Fixed many tests to correct for intended vs observed behavior

Arlodotexe committed:
Revert "Remove Minimum/Maximum Epsilon delta property and usage, fixed issue where Maximum values less than Minimum would erroneously overwrite Minimum"

This reverts commit c28bc00151f247142eda16cce5b9cc82249f3b6b.

Arlodotexe committed:
Remove exception for equal Minimum and Maximum values in RangeSelector

Arlodotexe committed:
Restore missing Epsilon field from revert

Arlodotexe committed:
Merge branch 'main' into dwm

michael-hawker commented:
Finally actually got a clean test-pass, but @Arlodotexe this still causes test failures:

```
Failed Initialize_MinEqMax (0,0,0,0) [198 ms]
```

Let's just fix this proper in 8.3, as it seems it's a mess right now to fix this.
